src/client/options/mod.rs

```diff
@@ -834,6 +834,19 @@ fn validate_userinfo(s: &str, userinfo_type: &str) -> Result<()> {
         }
         .into());
     }
+
+    // All instances of '%' in the username must be part of an percent-encoded substring. This means
```

Comment:
Previously, we had a check that would error if a percent character was present in the decoded username from the URI. This let us pass [this connection string spec test](https://github.com/mongodb/specifications/blob/4f89b79b6be51d6a61ae4196d07a0d6940d80be5/source/connection-string/tests/invalid-uris.yml#L236). However, [this auth spec test](https://github.com/mongodb/specifications/blob/4f89b79b6be51d6a61ae4196d07a0d6940d80be5/source/auth/tests/connection-string.yml#L348) contains an encoded percent sign in a valid URI. The restriction given in the [connection string spec](https://github.com/mongodb/specifications/blob/master/source/connection-string/connection-string-spec.rst#userinfo-optional) states that "If the user information contains an at-sign ("@"), more than one colon (":"), or a percent-sign ("%") that does not match the rules for "pct-encoded", then an exception MUST be thrown informing the user that the username and password must be URL encoded", but says nothing about disallowing an encoded percent sign, so I removed the previous check and added one here that checks for any `%` that isn't followed by two hex digits.

Reply:
Will `percent_encoding::percent_decode_str` not error already if an unescaped "%" is included in the input? We call this in our `percent_decode` function which is invoked right after `validate_userinfo` is.

Reply:
No, I don't think it will: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ef5d67a3353011e4bb59b98e568b9861